kafka/consumer.go

```diff
@@ -0,0 +1,393 @@
+		}
+	}
+
+	for _, part := range c.Partitions {
+		_, offset, err := c.consumer.QueryWatermarkOffsets(c.conf.Topics[0], part, c.conf.MetadataTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get newest offset for topic %s part %d: %s", c.conf.Topics[0], part, err)
+		}
+		c.bootTimeOffsets[part] = offset
+		c.partitionOffset[part] = stats.NewGauge64(fmt.Sprintf("%s.%d.offset", c.conf.GaugePrefix, part))
+		c.partitionLogSize[part] = stats.NewGauge64(fmt.Sprintf("%s.%d.log_size", c.conf.GaugePrefix, part))
+		c.partitionLag[part] = stats.NewGauge64(fmt.Sprintf("%s.%d.lag", c.conf.GaugePrefix, part))
+	}
+
+	return &c, nil
+}
+
+// Creates a lag monitor for the given size
+// This needs to be called before Start() or StartAndAwaitBacklog() to prevent
+// race conditions between initializing the lag monitor and setting lag values
+func (c *Consumer) InitLagMonitor(size int) {
+	c.LagMonitor = NewLagMonitor(size, c.Partitions)
+}
+
+func (c *Consumer) Start(processBacklog *sync.WaitGroup) error {
+	err := c.startConsumer()
+	if err != nil {
+		return fmt.Errorf("Failed to start consumer: %s", err)
+	}
+
+	go c.monitorLag(processBacklog)
+
+	for _ = range c.Partitions {
+		go c.consume()
+	}
+
+	return nil
+}
+
+func (c *Consumer) StartAndAwaitBacklog(backlogProcessTimeout time.Duration) error {
+	pre := time.Now()
+	processBacklog := new(sync.WaitGroup)
+	processBacklog.Add(len(c.Partitions))
+
+	err := c.Start(processBacklog)
+	if err != nil {
+		return err
+	}
+
+	// wait for our backlog to be processed before returning.  This will block metrictank from consuming metrics until
+	// we have processed old metricPersist messages. The end result is that we wont overwrite chunks in cassandra that
+	// have already been previously written.
+	// We don't wait more than backlogProcessTimeout for the backlog to be processed.
+	log.Info("kafka-consumer: Waiting for metricPersist backlog to be processed.")
+	backlogProcessed := make(chan struct{}, 1)
+	go func() {
+		processBacklog.Wait()
+		backlogProcessed <- struct{}{}
+	}()
+
+	select {
+	case <-time.After(backlogProcessTimeout):
+		log.Warn("kafka-consumer: Processing metricPersist backlog has taken too long, giving up lock after %s.", backlogProcessTimeout)
+	case <-backlogProcessed:
+		log.Info("kafka-consumer: MetricPersist backlog processed in %s.", time.Since(pre))
+	}
+
+	return nil
+}
+
+func (c *Consumer) consume() {
+	c.wg.Add(1)
+	defer c.wg.Done()
+
+	var ok bool
+	var offsetPtr *int64
+	events := c.consumer.Events()
+	for {
+		select {
+		case ev := <-events:
+			switch e := ev.(type) {
+			case confluent.AssignedPartitions:
+				c.consumer.Assign(e.Partitions)
+				log.Info("kafka-consumer: Assigned partitions: %+v", e)
+			case confluent.RevokedPartitions:
+				c.consumer.Unassign()
+				log.Info("kafka-consumer: Revoked partitions: %+v", e)
+			case confluent.PartitionEOF:
+				fmt.Printf("%% Reached %v\n", e)
```

Comment:
this should either be removed, or use `log.Debug`.  But i think we should just set `enable.partition.eof` to `false` in the confluent.ConfigMap to prevent these events from being emitted.